std-bits/table/src/main/java/org/enso/table/write/DelimitedWriter.java

```diff
@@ -0,0 +1,191 @@
+    } else {
+      this.quote = null;
+      this.quoteChar = '\0';
+    }
+
+    if (quoteEscape != null) {
+      if (quoteEscape.isEmpty()) {
+        throw new IllegalArgumentException(
+            "Empty quote escapes are not supported. Set the escape to `Nothing` to disable escaping quotes.");
+      }
+      if (quoteEscape.length() > 1) {
+        throw new IllegalArgumentException(
+            "Quote escapes consisting of multiple characters or code units are not supported.");
+      }
+
+      this.quoteEscape = quoteEscape;
+      this.quoteEscapeChar = quoteEscape.charAt(0);
+    } else {
+      this.quoteEscape = null;
+      this.quoteEscapeChar = '\0';
+    }
+
+    if (this.quoteEscape == this.quote) {
+      quoteReplacement = this.quote + "" + this.quote;
+      quoteEscapeReplacement = null;
+    } else {
+      quoteReplacement = this.quoteEscape + "" + this.quote;
+      quoteEscapeReplacement = this.quoteEscape + "" + this.quoteEscape;
+    }
+
+    this.writeQuoteBehavior = writeQuoteBehavior;
+    this.writeHeaders = writeHeaders;
+    emptyValue = this.quote + "" + this.quote;
+  }
+
+  public void write(Table table) throws IOException {
+    int numberOfColumns = table.getColumns().length;
+    assert numberOfColumns == columnFormatters.length;
+
+    if (writeHeaders) {
+      boolean quoteAllHeaders = writeQuoteBehavior == WriteQuoteBehavior.ALWAYS;
+      for (int col = 0; col < numberOfColumns; ++col) {
+        boolean isLast = col == numberOfColumns - 1;
+        writeCell(table.getColumns()[col].getName(), isLast, quoteAllHeaders);
+      }
+    }
+
+    int numberOfRows = table.rowCount();
+    for (int row = 0; row < numberOfRows; ++row) {
+      for (int col = 0; col < numberOfColumns; ++col) {
+        boolean isLast = col == numberOfColumns - 1;
+        Object cellValue = table.getColumns()[col].getStorage().getItemBoxed(row);
+        String formatted = columnFormatters[col].format(cellValue);
+        boolean wantsQuoting =
+            writeQuoteBehavior == WriteQuoteBehavior.ALWAYS && wantsQuotesInAlwaysMode(cellValue);
+        writeCell(formatted, isLast, wantsQuoting);
+      }
+    }
+
+    output.flush();
+  }
+
+  private boolean wantsQuotesInAlwaysMode(Object value) {
+    return value instanceof String || !isNonTextPrimitive(value);
+  }
+
+  private boolean isNonTextPrimitive(Object value) {
+    return value instanceof Long
+        || value instanceof Double
+        || value instanceof Boolean
+        || value instanceof LocalDate
+        || value instanceof LocalDateTime
+        || value instanceof LocalTime
+        || value instanceof ZonedDateTime;
+  }
+
+  private boolean quotingEnabled() {
+    return writeQuoteBehavior != WriteQuoteBehavior.NEVER;
+  }
+
+  private void writeCell(String value, boolean isLastInRow, boolean wantsQuoting)
+      throws IOException {
+    String processed = value == null ? "" : quotingEnabled() ? quote(value, wantsQuoting) : value;
```

Comment:
```suggestion
    String processed = value == null ? "" : wantsQuoting && quotingEnabled() ? quote(value) : value;
```
Then can simplify `quote`

Reply:
This won't work, because even if `wantsQuoting` is `false`, we may still quote in case that the value contains the delimiter or quote character.

Maybe I should modify the variable name - the intent of `wantsQuoting` is "does this value want to be 'always' quoted".